### ISSUE TITLE
Utterance Queue is not emptying wih the Both Hands interaction

### ISSUE THREAD

**Issue by terracoda** [phetsims/ratio-and-proportion#214]
This might be related to #200, but I don't think the utterance queue is clearing out as as it should.

I get tonnes of response pile-ups while using both hands, and just now, silenced voice over with the Control key, Tabbed to the Tick Mark Options radio buttons and the Both Hands started up again.

There is something sticky going on with the utterances. They stick around way too long.

I uploaded a scree capture here
https://drive.google.com/file/d/1wxzkoSagyFadfT_w7upSvdzS_0YNb4Wo/view?usp=sharing

**Comment by terracoda** [comment]
@zepumph, the reason I am thinking that this might be related to #200 is that it might be possible that the utterance queue doesn't know the interaction stopped because there is no explicit release, so then VO just continues with the queued responses once VO is reactivated.

**Comment by zepumph** [comment]
I feel like the main tool that we have in this case is a parameter on each Utterance that controls how long before the utterance is spoken. During this delay, if a new alert from the same utterance comes in, the timing restarts. I feel like it is quite likely that you are familiar with this parameter. 


NVDA sounds quite good in general with the current parameters, so my first thought would be to have you take a stab at tweaking these params to see if any other values sound better on your end.

Query parameters are like, and these are the defaults:
https://bayes.colorado.edu/dev/phettest/ratio-and-proportion/ratio-and-proportion_en.html?brand=phet&ea&bothHandsContextDelay=1000&bothHandsObjectDelay=500

Documentation at 
https://github.com/phetsims/ratio-and-proportion/blob/master/js/common/RAPQueryParameters.js#L54-L76

**Comment by zepumph** [comment]
@terracoda said today at design meeting, this is not a matter of tweaking, but instead this is a real issue, with way too many  pileups that aren't clearing out. @jessegreenberg, it may be nice to discuss this with you.

**Comment by zepumph** [comment]
@jessegreenberg was a huge help on this today. Together we were able to reproduce the issue on Mac. We found that the problem doesn't like within the UtteranceQueue parameters, because this pile up is happening after the queue has been drained, so this is a screen-reader issue. 

We also found that the behavior seems to be correct as it pertains to the specification of `aria-live="polite"`. VO let's the first alert completely finish before moving onto the next (https://stackoverflow.com/a/27558738/3408502 was helpful to look at for a refresher). Next we noted that this isn't a problem with the object/context responses with the mass sliders in GFL. We saw that because the object responses were implemented with aria-valuetext, each new setting of aria-valuetext would cut into the alerts and clear the pile up. We found that for this specific case, it worked pretty well to have the object response use an "assertive" aria-live alert instead of a "polite" one. Up until this point, 100% of all our alerts have been polite. 

Above implements the capability/option to make an Utterance alert assertive when desired. 

@terracoda please try out master, and see if the pile ups now clear whenever a new object response is alerted.

**Comment by zepumph** [comment]
For me to review:
- [ ] I'm surprised the a11y view's orange section is working. I wonder why adding assertive elements doesn't change the view? Maybe we should highlight the assertive ones for clarity?

**Comment by terracoda** [comment]
Nice work sorting out the difference between aria-valuetext and aria-live regions.

I think you are into something.

Is possible to flush the queue after an assertive response or something like that because, now I can get multiple polite responses after the latest assertive response.

Or can you make it so they are both assertive and I'll try that?

**Comment by terracoda** [comment]
Wait a second, it sound much better than yesterday.

**Comment by terracoda** [comment]
@zepumph, I agree, it would be good to highlight the assertive responses in the A11y View.

**Comment by terracoda** [comment]
BTW @jessegreenberg and @zepumph, great detective work here figuring out differences between `aria-valuetext` and `aria-live regions` for object responses.

So, the responses are working much better. If I move the hands relatively quickly, it seems the polite responses about proximity to ratio do not pile up. I just get "At challenge ratio", at the end when I hear the chime.

When I move them more slowly, the polite responses pile-up a bit, I can easily get 3 proximity-to-ratio responses firing one after the other after the hands are at the challenge, after I heard the chime and after the last object response about distance and position.

Perhaps, here we can now play with the timing and make it so the ratio responses do not stick around too long.

@zepumph, will the parameters you created still work? Please advise.

**Comment by zepumph** [comment]
>I'm surprised the a11y view's orange section is working. I wonder why adding assertive elements doesn't change the view?

It looks like we are using `mutation.addedNodes`, and so even though we are now adding an observer onto the parent div of all the polite elements (and another one for all the assertive ones), the addedNode is still the child `p` tag with the alert. I think this is fine for now, but could be improved. I'll make a note over in https://github.com/phetsims/chipper/issues/991.

**Comment by zepumph** [comment]
>@zepumph, will the parameters you created still work? Please advise.

Yes, it sounds like you are totally right. It's tweaking time! 

From what you said:

>When I move them more slowly, the polite responses pile-up a bit, I can easily get 3 proximity-to-ratio responses firing one after the other after the hands are at the challenge, after I heard the chime and after the last object response about distance and position.

It sounds like the first place to start is with decreasing the time on the context alert. Another way to potentially do the same thing is by increasing the time on the object response. It's just a matter of how quickly we should be getting these alerts. 

Try them both out and see! For context, the sliders we have in Ohms Law use 1000 for their value changes (which are that sim's context response).

**Comment by terracoda** [comment]
@zepumph, just noting that when using the sliders on VoiceOver, all the context responses containing distance from other hand and proximity to ratio pile up easily, too.

**Comment by zepumph** [comment]
> @zepumph, just noting that when using the sliders on VoiceOver, all the context responses containing distance from other hand and proximity to ratio pile up easily, too.

Thanks, I made https://github.com/phetsims/ratio-and-proportion/issues/220 because that is actually a totally different code path. I can see why it is buggy.

**Comment by terracoda** [comment]
object responses are assertive
context responses are polite

**Comment by terracoda** [comment]
Discussed at meeting.
For now MK is going to change the context response delay to 2 seconds.
Ideally we should design shorter object responses. We'll wait until after interviews and after Lock Ratio is designed.

**Comment by zepumph** [comment]
Today @terracoda, @Matthew-Moore240 and I thought through this problem a bit more. We found that the issue is with polite alerts ready to be spoken by the screen reader when an assertive alert comes in and has priority. In NVDA, there is internal behavior that clears all queued polite alerts when an assertive alert comes through. On VoiceOver this doesn't happen, thus it is possible to get a build up of context responses after the assertive alerts go through. Here is a diagram that helped me understand it:

```




                                           time where voice over gets the 
context response delay      context response but doesn't use it, so there is pile up
|----------------------------|==========|       

time for screen reader to read object response
|----------------------------------------------|

```

The only way to guarantee reproduceable experiences is to know how quickly a screen reader will speak the object response. This is an impossibility (of course). As @terracoda mentioned, after brainstorming about other potential solutions, we feel like the most robust way to solve this is to make the object response shorter. This may prove difficult, and we don't want to open that box at the moment (at least before interviews). Thus for now I will update the context response delay to be a bit longer, which helped a lot when @terracoda was trying it out.

**Comment by zepumph** [comment]
@terracoda, I think that this is ready for interviews at this point. If you agree, please remove this issue from the milestone (as we may come back for more tweaking later.

**Comment by terracoda** [comment]
@zepumph, think look good in the A11y View, but I can't load the sim for testing. I am getting this error:
![Screen Shot 2020-10-30 at 10 48 55 AM](https://user-images.githubusercontent.com/1902116/97709798-dc1ce180-1a9d-11eb-9430-975f61ea1335.png)

Can you have a look?

**Comment by terracoda** [comment]
I generally use PhET Marks locally at local host.
I always pull before I test and I often clone missing repos, too.
I always empty caches and clear history, too.

I don't know why for this sim it has been that I often can't load it and not always for the same reason.

I can't load it no locally after pulling, cloning, emptying caches, and clearing history.
<img width="885" alt="Screen Shot 2020-11-04 at 11 42 10 AM" src="https://user-images.githubusercontent.com/1902116/98128852-ff2b0500-1e92-11eb-8ce2-a6bcd1ed74ad.png">

**Comment by terracoda** [comment]
I'll try phettest directly.

**Comment by terracoda** [comment]
I can load from phettest directly.

**Comment by terracoda** [comment]
I verified the both hands interaction.
It was working as intended, so closing this issue.
